proposals/0000-asset-primitive.md

```diff
@@ -0,0 +1,233 @@
+
+All of these approaches work today (Astro v0.21), but suffer from an important problem: they resolve assets at **run-time** (aka "during the render"). For example, a component like `<Icon name="book">` can't know the value for `props.name` until the component and page itself is being run and rendered. This is even better illustrated if the value is a variable, like `<Icon name={fetchedAPIUserData.avatarIcon}>`. Astro would never know what this is.
+
+Why is this is a problem? To be clear, it's not a problem *today* because run-time and build-time are the same. In Astro today, **run-time** and **build-time** both happen during your build, ahead-of-time, before the site is deployed to the server. Runtime has full access to your machine, and every source file in your repo.
+
+However, with SSR support coming this will no longer always be true. If `<Icon name={fetchedAPIUserData.avatarIcon}>` renders for the first time on the server (ex: on Netlify, on Vercel) then Astro has no idea which icon you need to include in your build at build-time, which is when the decision of "what to include in your build" is made.
+
+This RFC is a priority to support our community members who are building components and integrations that use assets, and who need an asset strategy that we can commit to working in the future.  There have been different ideas thrown around to solve for this, some of which I will touch on in the "Alternatives" section of this RFC.
+
+### Goals
+
+This proposal outlines an **asset primitive** that accomplishes the following goals:
+
+- Works for Fonts, Icons, and Images.
+- Can reasonably extend to other asset types in the future.
+- Can reasonably extend to support image optimizations in the future.
+- Works with SSR, where all assets must be known at build-time, with 100% accuracy.
+- Works across frameworks (`.jsx`, `.svelte`, `.vue`). Not an `.astro`-only solution.
+- Introduces an acceptable amount of boilerplate (as little as possible, or none).
+- Can power custom 3rd-party integrations and components.
+- Can power future, higher-level Astro core features.
+
+# Detailed design
+
+## The Asset Primitive
+
+```js
+import imageBook from '../assets/book.png';
+  // imageBook is type: ImageAsset (.png|.jpg|.jpeg|.webp)
+  // {url, width, height, blurHashUrl}
+import iconCookie from '../assets/cookie.svg';
+  // iconCookieis type: SvgAsset (.svg)
+  // {width, height, content}
+import robotoFont400 from '../assets/fonts/roboto-slab-v22-latin-regular.woff2';
+  // robotoFont400 is type: string
+  // [If no custom asset is defined, Astro will default to including the asset
+  // in your build and returning a single URL as the result. This is Vite default behavior.]
+```
+
+- The asset primitive is imported via ESM.
+- Astro will define two custom `Asset` types: `ImageAsset` and `SvgAsset`
+- All other assets will just return a URL path string to the asset, and include that asset in the final build.
```

Comment:
Is there a definitive list of all things that could be called an asset?

This line says that:

> All other assets will just return a URL path

But it doesn't say what these 2 special asset types will return.

Reply:
clarified

Reply:
Clarified that this info exists in the code snippet at the top of the section